internal/factsengine/factsengine.go

```diff
@@ -0,0 +1,174 @@
+
+	return nil
+}
+
+func (c *factsEngine) Listen(ctx context.Context) {
+	log.Infof("Listening for facts gathering events...")
+	defer c.Unsubscribe()
+
+	err := c.factsServiceAdapter.Listen(c.agentID, c.handleRequest)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *factsEngine) handleRequest(request []byte) error {
+	factsRequests, err := parseFactsRequest(request)
+	if err != nil {
+		log.Errorf("Invalid facts request: %s", err)
+		return err
+	}
+
+	gatheredFacts, err := gatherFacts(factsRequests, c.factGatherers)
+	if err != nil {
+		log.Errorf("Error gathering facts: %s", err)
+		return err
+	}
+
+	if err := c.publishFacts(gatheredFacts); err != nil {
+		log.Errorf("Error publishing facts: %s", err)
+		return err
+	}
+
+	return nil
+}
+
+func gatherFacts(groupedFactsRequest *gatherers.GroupedFactsRequest, factGatherers map[string]gatherers.FactGatherer) (*gatherers.FactsResult, error) {
+	factsResults := &gatherers.FactsResult{
+		ExecutionID: groupedFactsRequest.ExecutionID,
+	}
+	log.Infof("Starting facts gathering process")
+
+	// Gather facts asynchronously
+	var wg sync.WaitGroup
+
+	for gathererType, factsRequest := range groupedFactsRequest.Facts {
+
+		g, exists := factGatherers[gathererType]
+		if !exists {
+			log.Errorf("Fact gatherer %s does not exist", gathererType)
+			continue
+		}
+
+		// Execute the fact gathering asynchronously and in parallel
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, factRequest []*gatherers.FactRequest) {
+			defer wg.Done()
+			newFacts, err := g.Gather(factRequest)
+			if err == nil {
+				factsResults.Facts = append(factsResults.Facts, newFacts[:]...)
+			} else {
+				log.Error(err)
+			}
+		}(&wg, factsRequest)
+	}
+
+	wg.Wait()
+
+	log.Infof("Requested facts gathered")
+	return factsResults, nil
+}
+
+func parseFactsRequest(request []byte) (*gatherers.GroupedFactsRequest, error) {
+	var factsRequest gatherers.FactsRequest
+	var groupedFactsRequest *gatherers.GroupedFactsRequest
+
+	err := json.Unmarshal(request, &factsRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	groupedFactsRequest = &gatherers.GroupedFactsRequest{
+		ExecutionID: factsRequest.ExecutionID,
+		Facts:       make(map[string][]*gatherers.FactRequest),
+	}
+
+	// Group the received facts by gatherer type, so they are executed in the same moment with the same source of truth
+	for _, factRequest := range factsRequest.Facts {
+		groupedFactsRequest.Facts[factRequest.Gatherer] = append(groupedFactsRequest.Facts[factRequest.Gatherer], factRequest)
+	}
+
+	return groupedFactsRequest, nil
+}
+
+func buildResponse(facts *gatherers.FactsResult) ([]byte, error) {
+	log.Infof("Building gathered facts response...")
+
+	jsonFacts, err := json.Marshal(facts)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("Gathered facts response built properly")
+
+	return jsonFacts, nil
+}
+
+func prettyString(str []byte) string {
+	var prettyJSON bytes.Buffer
+	json.Indent(&prettyJSON, str, "", "  ")
```

Comment:
we should check the error here from Indent

Reply:
Done